### PR TITLE
Flywheel

### PR DESCRIPTION
**Pull request by nindroz**
- created flywheel subsystem
- wrote function set flywheel with feedforward and PID

### CODE REVIEW

File: 2021-offseason/src/main/java/frc/robot/RobotContainer.java [hunk 1]
```diff
@@ -22,6 +25,7 @@
 public class RobotContainer {
   // The robot's subsystems and commands are defined here...
   private final ExampleSubsystem m_exampleSubsystem = new ExampleSubsystem();
+  private final Flywheel m_flywheel = new Flywheel();
```
Comment (uniqueusername) [inline 2021-offseason/src/main/java/frc/robot/RobotContainer.java:28]: ``s_flywheel``

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import com.revrobotics.CANSparkMax;
+
+import edu.wpi.first.wpilibj.controller.ProfiledPIDController;
+import edu.wpi.first.wpilibj.controller.SimpleMotorFeedforward;
+import edu.wpi.first.wpilibj.trajectory.TrapezoidProfile;
+import frc.robot.Constants;
+
+public class Flywheel extends SubsystemBase {
+  
+  //fields
```
Comment (uniqueusername) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:18]: ``// fields``

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  @Override
+  public void simulationPeriodic() {
+    // This method will be called once per scheduler run during simulation
+  }
+
+  //will automatically be called periodically
+  protected void useOutput(double output, TrapezoidProfile.State setpoint){
+    double fForward = feedForward.calculate(setpoint.position, setpoint.velocity)/12;
```
Comment (uniqueusername) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:72]: this variable could be better named; what is the difference between ``feedForward`` and ``fForward``?

**Review by uniqueusername — CHANGES_REQUESTED** [review]

File: 2021-offseason/src/main/java/frc/robot/Constants.java [hunk 4]
```diff
@@ -4,13 +4,17 @@
...
+ * The Constants class provides a convenient place for teams to hold robot-wide
+ * numerical or boolean constants. This class should not be used for any other
+ * purpose. All constants should be declared globally (i.e. public static). Do
+ * not put anything functional in this class.
  *
- * <p>It is advised to statically import this class (or one of its inner classes) wherever the
- * constants are needed, to reduce verbosity.
+ * <p>
+ * It is advised to statically import this class (or one of its inner classes)
+ * wherever the constants are needed, to reduce verbosity.
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/Constants.java:17]: why were these changed?

File: 2021-offseason/src/main/java/frc/robot/Constants.java [hunk 5]
```diff
@@ -34,4 +38,31 @@
         public static final int DRIVE_CONTROLLER = 0;
         public static final int OPERATOR_CONTROLLER = 1;
     }
+
+    public static final class kFlywheel {
+
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/Constants.java:43]: nit: remove extra newline
also line 64

Comment (KTOmega) [reply 2021-offseason/src/main/java/frc/robot/Constants.java:43]: and 66/67

File: 2021-offseason/src/main/java/frc/robot/commands/ExampleCommand.java [hunk 6]
```diff
@@ -34,7 +34,6 @@ public void execute() {}
   // Called once the command ends or is interrupted.
   @Override
   public void end(boolean interrupted) {}
-
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/commands/ExampleCommand.java:37]: revert this change, this isn't relevant to this pr

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+import com.revrobotics.CANSparkMax;
+
+import edu.wpi.first.wpilibj.controller.ProfiledPIDController;
+import edu.wpi.first.wpilibj.controller.SimpleMotorFeedforward;
+import edu.wpi.first.wpilibj.trajectory.TrapezoidProfile;
+import frc.robot.Constants;
+
+public class Flywheel extends SubsystemBase {
+
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:17]: nit: style: remove extra newline

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+public class Flywheel extends SubsystemBase {
+  
+  //fields
+  private final CANSparkMax flywheelMain;
+  private final CANSparkMax flywheelFollower;
+  private final ProfiledPIDController pidController;
+  private final SimpleMotorFeedforward feedForward;
+
+  public Flywheel() {
+
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:25]: nit: style: remove extra newline

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    pidController = new ProfiledPIDController(
+    Constants.kFlywheel.kP,
+    Constants.kFlywheel.kI,
+    Constants.kFlywheel.kD,
+    new TrapezoidProfile.Constraints(
+        Constants.kFlywheel.MAX_VELOCITY,
+        Constants.kFlywheel.MAX_ACCELERATION
+      )
+    );
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:37]: nit: style: better indentation
```suggestion
    pidController = new ProfiledPIDController(
      Constants.kFlywheel.kP,
      Constants.kFlywheel.kI,
      Constants.kFlywheel.kD,
      new TrapezoidProfile.Constraints(
        Constants.kFlywheel.MAX_VELOCITY,
        Constants.kFlywheel.MAX_ACCELERATION
      )
    );
```

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+    feedForward = new SimpleMotorFeedforward(
+      Constants.kFlywheel.kS,
+      Constants.kFlywheel.kV,
+      Constants.kFlywheel.kA
+    );
+    
+    //basic configuring
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:45]: nit: style: space after start of comment
```suggestion
    // basic configuring
```

Comment (KTOmega) [reply 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:45]: ary u missed this one

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    flywheelFollower.restoreFactoryDefaults();
+    
+    flywheelMain.setSmartCurrentLimit(Constants.kFlywheel.CURRENT_LIMIT);
+    flywheelFollower.setSmartCurrentLimit(Constants.kFlywheel.CURRENT_LIMIT);
+
+    flywheelMain.setInverted(Constants.kFlywheel.MAIN_INVERTED);
+    flywheelFollower.setInverted(Constants.kFlywheel.FOLLOWER_INVERTED);
+    flywheelFollower.follow(flywheelMain, true);
+
+    pidController.setTolerance(0,Constants.kFlywheel.ERR_TOLERANCE);
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:56]: nit: style: space after comma
```suggestion
    pidController.setTolerance(0, Constants.kFlywheel.ERR_TOLERANCE);
```

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:57]: nit: style: remove extra newline

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+  public void periodic() {
+    // This method will be called once per scheduler run
+  }
+
+  @Override
+  public void simulationPeriodic() {
+    // This method will be called once per scheduler run during simulation
+  }
+
+  //will automatically be called periodically
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:70]:
```suggestion
  // will automatically be called periodically
```

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+  protected void useOutput(double output, TrapezoidProfile.State setpoint){
+    double fForward = feedForward.calculate(setpoint.position, setpoint.velocity)/12;
+    flywheelMain.set(output+fForward);
+
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:74]: nit: style: remove extra newline

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  protected double getMeasurement(){
+    return flywheelMain.getEncoder().getVelocity();
+  }
+
+  public void setGoal(double goal){
+    pidController.setGoal(goal);
+  }
+
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:84]: nit: style: remove extra newline

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+    // This method will be called once per scheduler run
+  }
+
+  @Override
+  public void simulationPeriodic() {
+    // This method will be called once per scheduler run during simulation
+  }
+
+  //will automatically be called periodically
+  protected void useOutput(double output, TrapezoidProfile.State setpoint){
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:71]: nit: style: space before curly brace
(also check your other methods too)

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    double fForward = feedForward.calculate(setpoint.position, setpoint.velocity)/12;
+    flywheelMain.set(output+fForward);
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:73]: nit: style: space around operator
```suggestion
    flywheelMain.set(output + fForward);
```

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+    //basic configuring
+    flywheelMain.restoreFactoryDefaults();
+    flywheelFollower.restoreFactoryDefaults();
+    
+    flywheelMain.setSmartCurrentLimit(Constants.kFlywheel.CURRENT_LIMIT);
+    flywheelFollower.setSmartCurrentLimit(Constants.kFlywheel.CURRENT_LIMIT);
+
+    flywheelMain.setInverted(Constants.kFlywheel.MAIN_INVERTED);
+    flywheelFollower.setInverted(Constants.kFlywheel.FOLLOWER_INVERTED);
+    flywheelFollower.follow(flywheelMain, true);
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:54]: maybe this boolean be pulled out as a constant?

File: 2021-offseason/src/main/java/frc/robot/RobotContainer.java (hunk 1 above)
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/RobotContainer.java:28]: member variable declared but not used

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java (hunk 3 above)
Comment (nindroz) [reply 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:72]: is fForwardCalculated better? or should I just change the name of the feedforward variable and call this that?

**Review by KTOmega — CHANGES_REQUESTED** [review]
mostly nitpicks about spacing, a few functional comments

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java (hunk 3 above)
Comment (KTOmega) [reply 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:72]: maybe `ffOutput`

File: 2021-offseason/src/main/java/frc/robot/Constants.java (hunk 4 above)
Comment (nindroz) [reply 2021-offseason/src/main/java/frc/robot/Constants.java:17]: might have just been me messing up spacing. Not sure why that would have merged though as I rebased main onto flywheel first

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  //will automatically be called periodically
+  protected void useOutput(double output, TrapezoidProfile.State setpoint){
+    double fForward = feedForward.calculate(setpoint.position, setpoint.velocity)/12;
+    flywheelMain.set(output+fForward);
+
+  }
+
+  protected double getMeasurement(){
```
Comment (jaredgoodman03) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:77]: use a more specific unit in your method names, just call it `getVelocity()` or something so it's less ambiguous

Comment (nindroz) [reply 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:77]: its an abstract method from the profiledPIDController class that I just defined

File: 2021-offseason/src/main/java/frc/robot/RobotContainer.java (hunk 1 above)
Comment (nindroz) [reply 2021-offseason/src/main/java/frc/robot/RobotContainer.java:28]: its for the future when I finish out this stuff after writing commands

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java (hunk 19 above)
Comment (jaredgoodman03) [reply 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:77]: understandable

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java (hunk 3 above)
Comment (nindroz) [reply 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:72]: oh thats smart, thanks!

File: 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java (hunk 18 above)
Comment (nindroz) [reply 2021-offseason/src/main/java/frc/robot/subsystems/Flywheel.java:54]: I think its always supposed to follow so i dont think its necessary

**Review by poroheart — APPROVED** [review]
😍😍😍
